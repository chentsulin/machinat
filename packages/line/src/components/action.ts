import { partSegment } from '@machinat/core/renderer';
import { PartSegment } from '@machinat/core/renderer/types';
import { annotateLineComponent } from '../utils';
import type { LineComponent } from '../types';

/**
 * @category Props
 */
type PostbackActionProps = {
  /** Label for the action */
  label: string;
  /**
   * String returned via webhook in the `postback.data` property of the postback
   * event. Max character limit: 300.
   */
  data: string;
  /**
   * Text displayed in the chat as a message sent by the user when the action is
   * performed. Required for quick reply buttons. Optional for the other message
   * types. Max character limit: 300.
   */
  displayText?: string;
  /** Alias of `displayText` */
  text?: string;
};

/** @internal */
const __PostbackAction = function PostbackAction(node, path) {
  const { label, data, text, displayText } = node.props;
  return [
    partSegment(node, path, {
      type: 'postback',
      data,
      label,
      displayText: displayText || text,
    }),
  ];
};
/**
 * When a control associated with this action is tapped, a postback event is
 * returned via webhook with the specified string in the data property.
 * @category Component
 * @props {@link PostbackActionProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#postback-action).
 */
export const PostbackAction: LineComponent<
  PostbackActionProps,
  PartSegment<any>
> = annotateLineComponent(__PostbackAction);

/**
 * @category Props
 */
type MessageActionProps = {
  /** Label for the action */
  label: string;
  /** Text sent when the action is performed. Max character limit: 300 */
  text?: string;
};

/** @internal */
const __MessageAction = function MessageAction(node, path) {
  const { label, text } = node.props;
  return [
    partSegment(node, path, {
      type: 'message',
      label,
      text,
    }),
  ];
};
/**
 * When a control associated with this action is tapped, the string in the
 * `text` property is sent as a message from the user.
 * @category Component
 * @props {@link MessageActionProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#message-action).
 */
export const MessageAction: LineComponent<
  MessageActionProps,
  PartSegment<any>
> = annotateLineComponent(__MessageAction);

/**
 * @category Props
 */
type UriActionProps = {
  /** Label for the action */
  label: string;
  /**
   * URI opened when the action is performed (Max character limit: 1000). The
   * available schemes are http, https, line, and tel.
   */
  uri: string;
};

/** @internal */
const __UriAction = function UriAction(node, path) {
  const { label, uri } = node.props;
  return [
    partSegment(node, path, {
      type: 'uri',
      label,
      uri,
    }),
  ];
};
/**
 * When a control associated with this action is tapped, the URI specified in
 * the `uri` property is opened.
 * @category Component
 * @props {@link UriActionProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#uri-action).
 */
export const UriAction: LineComponent<
  UriActionProps,
  PartSegment<any>
> = annotateLineComponent(__UriAction);

/** @internal */
const pad2 = (n) => (n < 10 ? `0${n}` : n);

/** @internal */
const fullDate = (d) =>
  `${d.getFullYear()}-${pad2(d.getMonth() + 1)}-${pad2(d.getDate())}`;

/** @internal */
const fullHourMinute = (d) => `${pad2(d.getHours())}:${pad2(d.getMinutes())}`;

/** @internal */
const dateToStringByMode = (mode, d) =>
  !(d instanceof Date)
    ? d
    : mode === 'datetime'
    ? `${fullDate(d)}T${fullHourMinute(d)}`
    : mode === 'date'
    ? fullDate(d)
    : fullHourMinute(d);

/**
 * @category Props
 */
type DateTimePickerActionProps = {
  /** Label for the action  */
  label: string;
  /**
   * String returned via webhook in the postback.data property of the postback
   * event. Max character limit: 300.
   */
  data: string;
  /** Action mode */
  mode: 'datetime' | 'date' | 'time';
  /** Initial value of date or time */
  initial?: string | Date;
  /**
   * Largest date or time value that can be selected. Must be greater than the
   * min value.
   */
  max?: string | Date;
  /**
   * Smallest date or time value that can be selected. Must be less than the max
   * value.
   */
  min?: string | Date;
};

/** @internal */
const __DateTimePickerAction = function DateTimePickerAction(node, path) {
  const { label, data, mode = 'datetime', initial, min, max } = node.props;

  return [
    partSegment(node, path, {
      type: 'datetimepicker',
      label,
      data,
      mode,
      initial: dateToStringByMode(mode, initial),
      max: dateToStringByMode(mode, max),
      min: dateToStringByMode(mode, min),
    }),
  ];
};
/**
 * When a control associated with this action is tapped, a postback event is
 * returned via webhook with the date and time selected by the user from the
 * date and time selection dialog. The datetime picker action does not support
 * time zones.
 * @category Component
 * @props {@link DateTimePickerActionProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#datetime-picker-action).
 */
export const DateTimePickerAction: LineComponent<
  DateTimePickerActionProps,
  PartSegment<any>
> = annotateLineComponent(__DateTimePickerAction);

/**
 * @category Props
 */
type CameraActionProps = {
  /** Label for the action. Max character limit: 20 */
  label: string;
};

/** @internal */
const __CameraAction = function CameraAction(node, path) {
  return [
    partSegment(node, path, {
      type: 'camera',
      label: node.props.label,
    }),
  ];
};
/**
 * This action can be configured only with quick reply buttons. When a button
 * associated with this action is tapped, the camera screen in LINE is opened.
 * @category Component
 * @props {@link CameraActionProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#camera-action).
 */
export const CameraAction: LineComponent<
  CameraActionProps,
  PartSegment<any>
> = annotateLineComponent(__CameraAction);

/**
 * @category Props
 */
type CameraRollActionProps = {
  /** Label for the action. Max character limit: 20 */
  label: string;
};

/** @internal */
const __CameraRollAction = function CameraRollAction(node, path) {
  return [
    partSegment(node, path, {
      type: 'cameraRoll',
      label: node.props.label,
    }),
  ];
};
/**
 * This action can be configured only with quick reply buttons. When a button
 * associated with this action is tapped, the camera roll screen in LINE is
 * opened.
 * @category Component
 * @props {@link CameraRollActionProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#camera-roll-action).
 */
export const CameraRollAction: LineComponent<
  CameraRollActionProps,
  PartSegment<any>
> = annotateLineComponent(__CameraRollAction);

/**
 * @category Props
 */
type LocationActionProps = {
  /** Label for the action. Max character limit: 20 */
  label: string;
};

/** @internal */
const __LocationAction = function LocationAction(node, path) {
  return [
    partSegment(node, path, {
      type: 'location',
      label: node.props.label,
    }),
  ];
};
/**
 * This action can be configured only with quick reply buttons. When a button
 * associated with this action is tapped, the location screen in LINE is opened.
 * @category Component
 * @props {@link LocationActionProps}
 * @guides Check official [reference](https://developers.line.biz/en/reference/messaging-api/#camera-roll-action).
 */
export const LocationAction: LineComponent<
  LocationActionProps,
  PartSegment<any>
> = annotateLineComponent(__LocationAction);

export type Action =
  | typeof PostbackAction
  | typeof MessageAction
  | typeof UriAction
  | typeof DateTimePickerAction
  | typeof CameraAction
  | typeof CameraRollAction
  | typeof LocationAction;
