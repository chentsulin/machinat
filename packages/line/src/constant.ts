export const CHANNEL_REQUEST_GETTER = Symbol(
  'channel_request_getter.line.machinat'
);
export const BULK_REQUEST_GETTER = Symbol('bulk_request_getter.line.machinat');

/** @internal */
export const LINE = 'line' as const;

/** @internal */
export const PATH_REPLY = 'v2/bot/message/reply';

/** @internal */
export const PATH_PUSH = 'v2/bot/message/push';

/** @internal */
export const PATH_MULTICAST = 'v2/bot/message/multicast';

/** @internal */
export const PATH_RICHMENU = 'v2/bot/richmenu';

export enum LiffContextOs {
  Ios,
  Android,
  Web,
}
