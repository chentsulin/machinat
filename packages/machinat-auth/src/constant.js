// @flow
import { namedInterface } from '@machinat/core/service';
import type { ServerAuthorizer, AuthModuleConfigs } from './types';

export const STATE_COOKIE_KEY = 'machinat_auth_state';
export const TOKEN_COOKIE_KEY = 'machinat_auth_token';
export const SIGNATURE_COOKIE_KEY = 'machinat_auth_signature';
export const ERROR_COOKIE_KEY = 'machinat_auth_error';

export const AUTH_SERVER_AUTHORIZERS_I = namedInterface<
  ServerAuthorizer<any, any>[]
>({
  name: 'ServerAuthorizersList',
  multi: true,
});

export const AUTH_MODULE_CONFIGS_I = namedInterface<AuthModuleConfigs>({
  name: 'AuthModuleConfigs',
});
