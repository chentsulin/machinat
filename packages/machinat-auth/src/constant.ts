import { makeInterface } from '@machinat/core/service';
import type { ServerAuthorizer, AuthModuleConfigs } from './types';

export const STATE_COOKIE_KEY = 'machinat_auth_state';
export const TOKEN_COOKIE_KEY = 'machinat_auth_token';
export const SIGNATURE_COOKIE_KEY = 'machinat_auth_signature';
export const ERROR_COOKIE_KEY = 'machinat_auth_error';

export const SERVER_AUTHORIZERS_I = makeInterface<ServerAuthorizer<any, any>[]>(
  { name: 'ServerAuthorizersList', multi: true }
);

export const MODULE_CONFIGS_I = makeInterface<AuthModuleConfigs>({
  name: 'AuthModuleConfigs',
});
