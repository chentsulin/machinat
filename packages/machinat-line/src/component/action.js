import { asPartComponent } from './utils';

const PostbackAction = ({ props: { label, data, text, displayText } }) => [
  {
    type: 'postback',
    data,
    label,
    displayText: displayText || text,
  },
];
const __PostbackAction = asPartComponent(PostbackAction);

const MessageAction = ({ props: { label, text } }) => [
  {
    type: 'message',
    label,
    text,
  },
];
const __MessageAction = asPartComponent(MessageAction);

const URIAction = ({ props: { label, uri } }) => [
  {
    type: 'uri',
    label,
    uri,
  },
];
const __URIAction = asPartComponent(URIAction);

const pad2 = n => (n < 10 ? `0${n}` : n);
const fullDate = d =>
  `${d.getFullYear()}-${pad2(d.getMonth() + 1)}-${pad2(d.getDate())}`;
const fullHourMinute = d => `${pad2(d.getHours())}:${pad2(d.getMinutes())}`;
const dateToStringByMode = (mode, d) =>
  !(d instanceof Date)
    ? d
    : mode === 'datetime'
    ? `${fullDate(d)}T${fullHourMinute(d)}`
    : mode === 'date'
    ? fullDate(d)
    : fullHourMinute(d);

const DateTimePickerAction = ({
  props: { label, data, mode, date, time, initial, min, max },
}) => {
  const modeRefined =
    mode || (!date === !time ? 'datetime' : date ? 'date' : 'time');

  return [
    {
      type: 'datetimepicker',
      label,
      data,
      mode: modeRefined,
      initial: dateToStringByMode(modeRefined, initial),
      max: dateToStringByMode(modeRefined, max),
      min: dateToStringByMode(modeRefined, min),
    },
  ];
};
const __DateTimePickerAction = asPartComponent(DateTimePickerAction);

const CameraAction = ({ props: { label } }) => [
  {
    type: 'camera',
    label,
  },
];
const __CameraAction = asPartComponent(CameraAction);

const CameraRollAction = ({ props: { label } }) => [
  {
    type: 'cameraRoll',
    label,
  },
];
const __CameraRollAction = asPartComponent(CameraRollAction);

const LocationAction = ({ props: { label } }) => [
  {
    type: 'location',
    label,
  },
];
const __LocationAction = asPartComponent(LocationAction);

export {
  __PostbackAction as PostbackAction,
  __MessageAction as MessageAction,
  __URIAction as URIAction,
  __DateTimePickerAction as DateTimePickerAction,
  __CameraAction as CameraAction,
  __CameraRollAction as CameraRollAction,
  __LocationAction as LocationAction,
};
