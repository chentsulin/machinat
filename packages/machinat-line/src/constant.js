// @flow
export const LINE = 'line';
export const LINE_NAMESPACE = 'Line';

export const LINE_NATIVE_TYPE = Symbol.for('machinat.line.native');
