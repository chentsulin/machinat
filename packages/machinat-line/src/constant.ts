export const LINE = 'line';

export const CHANNEL_API_CALL_GETTER = Symbol(
  'line.segment.channel_api_call_getter'
);
export const BULK_API_CALL_GETTER = Symbol('line.segment.bulk_api_call_getter');

export const PATH_REPLY = 'v2/bot/message/reply';
export const PATH_PUSH = 'v2/bot/message/push';
export const PATH_MULTICAST = 'v2/bot/message/multicast';
export const PATH_RICHMENU = 'v2/bot/richmenu';
