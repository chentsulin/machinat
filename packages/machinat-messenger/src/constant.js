// @flow
export const MESSENGER = 'messenger';
export const MESSENGER_NAMESPACE = 'Messenger';
export const MESSENGER_NATIVE_TYPE = Symbol.for('machinat.messenger.native');

export const ATTACHED_FILE_DATA = Symbol(
  'machinat.messenger.job.attached_file_data'
);

export const ATTACHED_FILE_INFO = Symbol(
  'machinat.messenger.job.attached_file_info'
);

export const ENTRY_MESSAGES = 'me/messages';

export const ENTRY_PASS_THREAD_CONTROL = 'me/pass_thread_control';
export const ENTRY_REQUEST_THREAD_CONTROL = 'me/request_thread_control';
export const ENTRY_TAKE_THREAD_CONTROL = 'me/take_thread_control';

export const ENTRY_MESSAGE_CREATIVES = 'me/message_creatives';
export const ENTRY_BROADCAST_MESSAGES = 'me/broadcast_messages';
export const ENTRY_CUSTOM_LABELS = 'me/custom_labels';
