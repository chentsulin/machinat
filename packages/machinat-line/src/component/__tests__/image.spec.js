import Machinat from 'machinat';
import { map } from 'machinat-utility';

import { LINE_NATIVE_TYPE } from '../../constant';
import {
  Image,
  Sticker,
  ImageMap,
  ImageMapVideoArea,
  ImageMapArea,
} from '../image';
import { URIAction, MessageAction } from '../action';
import renderHelper from './renderHelper';

const renderInner = (prop, route) =>
  map(prop, (node, path) => node.type(node, renderInner, path)[0], route) ||
  null;
const render = renderHelper(renderInner);

it.each(
  [Image, Sticker, ImageMap, ImageMapVideoArea, ImageMapArea].map(C => [
    C.name,
    C,
  ])
)('%s is a valid component', (_, Img) => {
  expect(Img.$$native).toBe(LINE_NATIVE_TYPE);
  expect(Img.$$getEntry).toBe(undefined);
});

it.each(
  [
    <Image url="https://..." previewURL="https://..." />,
    <Sticker packageId={1} stickerId={2} />,
    <ImageMap baseURL="https://..." alt="..." height={999}>
      <ImageMapArea
        label="foo"
        text="bar"
        x={123}
        y={456}
        width={654}
        height={321}
        action={<MessageAction label="foo" text="bar" />}
      />
      <ImageMapArea
        x={978}
        y={654}
        width={456}
        height={789}
        action={<URIAction label="foo" uri="https://..." />}
      />
    </ImageMap>,
    <ImageMap
      baseURL="https://..."
      alt="..."
      height={999}
      video={
        <ImageMapVideoArea
          url="https://..."
          previewURL="https://..."
          x={123}
          y={456}
          width={654}
          height={321}
        />
      }
    >
      <ImageMapArea
        label="foo"
        text="bar"
        x={123}
        y={456}
        width={654}
        height={321}
        action={<MessageAction label="foo" text="bar" />}
      />
    </ImageMap>,
    <ImageMap
      baseURL="https://..."
      alt="..."
      height={999}
      video={
        <ImageMapVideoArea
          url="https://..."
          previewURL="https://..."
          x={123}
          y={456}
          width={654}
          height={321}
          action={<URIAction label="foo" uri="https://..." />}
        />
      }
    >
      <ImageMapArea
        x={978}
        y={654}
        width={456}
        height={789}
        action={<URIAction label="foo" uri="https://..." />}
      />
    </ImageMap>,
  ].map(e => [e.type.name, e])
)('%s match snapshot', (_, element) => {
  const segments = render(element);
  expect(segments.length).toBe(1);

  const segment = segments[0];
  expect(segment.type).toBe('unit');
  expect(segment.node).toBe(element);
  expect(segment.path).toBe('$');
  expect(segment.value).toMatchSnapshot();
});
