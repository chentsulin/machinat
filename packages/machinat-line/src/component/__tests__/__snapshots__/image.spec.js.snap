// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image match snapshot 1`] = `
Object {
  "originalContentUrl": "https://...",
  "previewImageUrl": "https://...",
  "type": "image",
}
`;

exports[`ImageMap match snapshot 1`] = `
Object {
  "actions": Array [
    Object {
      "area": Object {
        "height": 321,
        "width": 654,
        "x": 123,
        "y": 456,
      },
      "label": "foo",
      "text": "bar",
      "type": "message",
    },
    Object {
      "area": Object {
        "height": 789,
        "width": 456,
        "x": 978,
        "y": 654,
      },
      "label": "foo",
      "linkUri": "https://...",
      "type": "uri",
    },
  ],
  "altText": "...",
  "baseSize": Object {
    "height": 999,
    "width": 1040,
  },
  "baseUrl": "https://...",
  "type": "imagemap",
  "video": undefined,
}
`;

exports[`ImageMap match snapshot 2`] = `
Object {
  "actions": Array [
    Object {
      "area": Object {
        "height": 321,
        "width": 654,
        "x": 123,
        "y": 456,
      },
      "label": "foo",
      "text": "bar",
      "type": "message",
    },
  ],
  "altText": "...",
  "baseSize": Object {
    "height": 999,
    "width": 1040,
  },
  "baseUrl": "https://...",
  "type": "imagemap",
  "video": Object {
    "area": Object {
      "height": 321,
      "width": 654,
      "x": 123,
      "y": 456,
    },
    "externalLink": undefined,
    "originalContentUrl": "https://...",
    "previewImageUrl": "https://...",
  },
}
`;

exports[`ImageMap match snapshot 3`] = `
Object {
  "actions": Array [
    Object {
      "area": Object {
        "height": 789,
        "width": 456,
        "x": 978,
        "y": 654,
      },
      "label": "foo",
      "linkUri": "https://...",
      "type": "uri",
    },
  ],
  "altText": "...",
  "baseSize": Object {
    "height": 999,
    "width": 1040,
  },
  "baseUrl": "https://...",
  "type": "imagemap",
  "video": Object {
    "area": Object {
      "height": 321,
      "width": 654,
      "x": 123,
      "y": 456,
    },
    "externalLink": Object {
      "label": "foo",
      "linkUri": "https://...",
    },
    "originalContentUrl": "https://...",
    "previewImageUrl": "https://...",
  },
}
`;

exports[`Sticker match snapshot 1`] = `
Object {
  "packageId": 1,
  "stickerId": 2,
  "type": "sticker",
}
`;
